crates/core/src/subscription/query.rs

```diff
@@ -866,27 +886,51 @@ mod tests {
-        CREATE TABLE EnemyState (entity_id BIGINT UNSIGNED, herd_id INTEGER, status INTEGER, type INTEGER, direction INTEGER);";
-        run(&db, &mut tx, sql_create, AuthCtx::for_testing())?;
+        // Create table [MobileEntityState]
+        let schema = &[
+            ("entity_id", AlgebraicType::U64),
+            ("location_x", AlgebraicType::I32),
+            ("location_z", AlgebraicType::I32),
+            ("destination_x", AlgebraicType::I32),
+            ("destination_z", AlgebraicType::I32),
+            ("is_running", AlgebraicType::Bool),
+            ("timestamp", AlgebraicType::U64),
+            ("dimension", AlgebraicType::U32),
+        ];
+        let indexes = &[(0, "entity_id"), (1, "location_x"), (2, "location_z")];
+        create_table(&db, &mut tx, "MobileEntityState", schema, indexes)?;
+
+        // Create table [EnemyState]
+        let schema = &[
+            ("entity_id", AlgebraicType::U64),
+            ("herd_id", AlgebraicType::I32),
+            ("status", AlgebraicType::I32),
+            ("type", AlgebraicType::I32),
+            ("direction", AlgebraicType::I32),
+        ];
+        let indexes = &[(0, "entity_id")];
+        create_table(&db, &mut tx, "EnemyState", schema, indexes)?;
 
-        let sql_create = "\
+        let sql_insert = "\
         insert into MobileEntityState (entity_id, location_x, location_z, destination_x, destination_z, is_running, timestamp, dimension) values (1, 96001, 96001, 96001, 1867045146, false, 17167179743690094247, 3926297397);\
         insert into MobileEntityState (entity_id, location_x, location_z, destination_x, destination_z, is_running, timestamp, dimension) values (2, 96001, 191000, 191000, 1560020888, true, 2947537077064292621, 445019304);
 
         insert into EnemyState (entity_id, herd_id, status, type, direction) values (1, 1181485940, 1633678837, 1158301365, 132191327);
         insert into EnemyState (entity_id, herd_id, status, type, direction) values (2, 2017368418, 194072456, 34423057, 1296770410);";
-        run(&db, &mut tx, sql_create, AuthCtx::for_testing())?;
-
-        let sql_query = "SELECT * FROM MobileEntityState JOIN EnemyState ON MobileEntityState.entity_id = EnemyState.entity_id WHERE location_x > 96000 AND MobileEntityState.location_x < 192000 AND MobileEntityState.location_z > 96000 AND MobileEntityState.location_z < 192000";
-        let q = compile_read_only_query(&db, &tx, &AuthCtx::for_testing(), sql_query)?;
-
-        for q in &q {
-            assert_eq!(
-                run_query(&db, &mut tx, q, AuthCtx::for_testing())?.len(),
-                1,
-                "Not return results"
-            );
+        run(&db, &mut tx, sql_insert, AuthCtx::for_testing())?;
+
+        let sql_query = "\
```

Comment:
This is not detected as an IndexJoin, despite actually creating the indexes above. It looks to be due to the first `query` element being an `IndexScan`, and the length thus being > 2. Not sure how to fix.

Reply:
Well yes, I thought about that. But the problem is something else. This is how the compiled query (with qualified wildcard) looks like; as you can see, `query` contains more than two elements, and the first one is `IndexScan`:

<details>
  <summary>Click to expand</summary>

```rust
QueryExpr {
    source: DbTable(
        DbTable {
            head: Header {
                table_name: "MobileEntityState",
                fields: [
                    Column {
                        field: Name {
                            table: "MobileEntityState",
                            field: "entity_id",
                        },
                        algebraic_type: Builtin(
                            U64,
                        ),
                    },
                    Column {
                        field: Name {
                            table: "MobileEntityState",
                            field: "location_x",
                        },
                        algebraic_type: Builtin(
                            I32,
                        ),
                    },
                    Column {
                        field: Name {
                            table: "MobileEntityState",
                            field: "location_z",
                        },
                        algebraic_type: Builtin(
                            I32,
                        ),
                    },
                    Column {
                        field: Name {
                            table: "MobileEntityState",
                            field: "destination_x",
                        },
                        algebraic_type: Builtin(
                            I32,
                        ),
                    },
                    Column {
                        field: Name {
                            table: "MobileEntityState",
                            field: "destination_z",
                        },
                        algebraic_type: Builtin(
                            I32,
                        ),
                    },
                    Column {
                        field: Name {
                            table: "MobileEntityState",
                            field: "is_running",
                        },
                        algebraic_type: Builtin(
                            Bool,
                        ),
                    },
                    Column {
                        field: Name {
                            table: "MobileEntityState",
                            field: "timestamp",
                        },
                        algebraic_type: Builtin(
                            U64,
                        ),
                    },
                    Column {
                        field: Name {
                            table: "MobileEntityState",
                            field: "dimension",
                        },
                        algebraic_type: Builtin(
                            U32,
                        ),
                    },
                ],
            },
            table_id: 6,
            table_type: User,
            table_access: Public,
        },
    ),
    query: [
        IndexScan(
            IndexScan {
                table: DbTable {
                    head: Header {
                        table_name: "MobileEntityState",
                        fields: [
                            Column {
                                field: Name {
                                    table: "MobileEntityState",
                                    field: "entity_id",
                                },
                                algebraic_type: Builtin(
                                    U64,
                                ),
                            },
                            Column {
                                field: Name {
                                    table: "MobileEntityState",
                                    field: "location_x",
                                },
                                algebraic_type: Builtin(
                                    I32,
                                ),
                            },
                            Column {
                                field: Name {
                                    table: "MobileEntityState",
                                    field: "location_z",
                                },
                                algebraic_type: Builtin(
                                    I32,
                                ),
                            },
                            Column {
                                field: Name {
                                    table: "MobileEntityState",
                                    field: "destination_x",
                                },
                                algebraic_type: Builtin(
                                    I32,
                                ),
                            },
                            Column {
                                field: Name {
                                    table: "MobileEntityState",
                                    field: "destination_z",
                                },
                                algebraic_type: Builtin(
                                    I32,
                                ),
                            },
                            Column {
                                field: Name {
                                    table: "MobileEntityState",
                                    field: "is_running",
                                },
                                algebraic_type: Builtin(
                                    Bool,
                                ),
                            },
                            Column {
                                field: Name {
                                    table: "MobileEntityState",
                                    field: "timestamp",
                                },
                                algebraic_type: Builtin(
                                    U64,
                                ),
                            },
                            Column {
                                field: Name {
                                    table: "MobileEntityState",
                                    field: "dimension",
                                },
                                algebraic_type: Builtin(
                                    U32,
                                ),
                            },
                        ],
                    },
                    table_id: 6,
                    table_type: User,
                    table_access: Public,
                },
                col_id: 1,
                lower_bound: Excluded(
                    I32(
                        96000,
                    ),
                ),
                upper_bound: Excluded(
                    I32(
                        192000,
                    ),
                ),
            },
        ),
        Select(
            Cmp {
                op: Logic(
                    And,
                ),
                lhs: Cmp {
                    op: Cmp(
                        Gt,
                    ),
                    lhs: Field(
                        Name(
                            Name {
                                table: "MobileEntityState",
                                field: "location_z",
                            },
                        ),
                    ),
                    rhs: Field(
                        Value(
                            I32(
                                96000,
                            ),
                        ),
                    ),
                },
                rhs: Cmp {
                    op: Cmp(
                        Lt,
                    ),
                    lhs: Field(
                        Name(
                            Name {
                                table: "MobileEntityState",
                                field: "location_z",
                            },
                        ),
                    ),
                    rhs: Field(
                        Value(
                            I32(
                                192000,
                            ),
                        ),
                    ),
                },
            },
        ),
        JoinInner(
            JoinExpr {
                rhs: QueryExpr {
                    source: DbTable(
                        DbTable {
                            head: Header {
                                table_name: "EnemyState",
                                fields: [
                                    Column {
                                        field: Name {
                                            table: "EnemyState",
                                            field: "entity_id",
                                        },
                                        algebraic_type: Builtin(
                                            U64,
                                        ),
                                    },
                                    Column {
                                        field: Name {
                                            table: "EnemyState",
                                            field: "herd_id",
                                        },
                                        algebraic_type: Builtin(
                                            I32,
                                        ),
                                    },
                                    Column {
                                        field: Name {
                                            table: "EnemyState",
                                            field: "status",
                                        },
                                        algebraic_type: Builtin(
                                            I32,
                                        ),
                                    },
                                    Column {
                                        field: Name {
                                            table: "EnemyState",
                                            field: "type",
                                        },
                                        algebraic_type: Builtin(
                                            I32,
                                        ),
                                    },
                                    Column {
                                        field: Name {
                                            table: "EnemyState",
                                            field: "direction",
                                        },
                                        algebraic_type: Builtin(
                                            I32,
                                        ),
                                    },
                                ],
                            },
                            table_id: 7,
                            table_type: User,
                            table_access: Public,
                        },
                    ),
                    query: [],
                },
                col_lhs: Name {
                    table: "MobileEntityState",
                    field: "entity_id",
                },
                col_rhs: Name {
                    table: "EnemyState",
                    field: "entity_id",
                },
            },
        ),
        Project(
            [
                Name(
                    Name {
                        table: "MobileEntityState",
                        field: "entity_id",
                    },
                ),
                Name(
                    Name {
                        table: "MobileEntityState",
                        field: "location_x",
                    },
                ),
                Name(
                    Name {
                        table: "MobileEntityState",
                        field: "location_z",
                    },
                ),
                Name(
                    Name {
                        table: "MobileEntityState",
                        field: "destination_x",
                    },
                ),
                Name(
                    Name {
                        table: "MobileEntityState",
                        field: "destination_z",
                    },
                ),
                Name(
                    Name {
                        table: "MobileEntityState",
                        field: "is_running",
                    },
                ),
                Name(
                    Name {
                        table: "MobileEntityState",
                        field: "timestamp",
                    },
                ),
                Name(
                    Name {
                        table: "MobileEntityState",
                        field: "dimension",
                    },
                ),
            ],
            Some(
                6,
            ),
        ),
    ],
}
```

</details>